File: apps/workbench/backend/api/v1/planning.py
from __future__ import annotations

from typing import Annotated, Any, Literal

from fastapi import APIRouter, Header, HTTPException, Request, Response, status
from pydantic import BaseModel, ConfigDict, Field

from ...application.case_service import CasePrincipal
from ...application.planning_service import (
    CompileDecisionSurfaceDraft,
    PlanningCheckpointDecisionDraft,
    PlanningService,
    PlanningServiceError,
    ReviseDecisionSurfaceDraft,
)
from sec_agent.workbench.api_contracts import request_trace_id


class CompileDecisionSurfaceCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_case_version: int = Field(ge=0)
    expected_summary_version: int = Field(ge=0)
    compiler_policy_ref: Literal["fixture:p36-three-cell-v1", "fixture:p36-ten-cell-v1"]
    pack_selection_ref: Literal["fixture:p36-ai-infrastructure-v1", "fixture:p36-ai-infrastructure-v2"]
    actor_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class DecisionSurfaceRevision(BaseModel):
    model_config = ConfigDict(extra="forbid")

    cell_id: str = Field(min_length=1)
    what_would_change: str = Field(min_length=1)
    stop_rule: str | None = Field(default=None, min_length=1)


class ReviseDecisionSurfaceCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_case_version: int = Field(ge=0)
    expected_decision_surface_contract_version: int = Field(ge=0)
    expected_checkpoint_version: int = Field(ge=0)
    changes: list[DecisionSurfaceRevision] = Field(min_length=1)
    actor_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class PlanningCheckpointDecisionCommand(BaseModel):
    model_config = ConfigDict(extra="forbid")

    decision: Literal["accept", "return"]
    expected_case_version: int = Field(ge=0)
    expected_decision_surface_contract_version: int = Field(ge=0)
    expected_checkpoint_version: int = Field(ge=0)
    actor_ref: str = Field(min_length=1)
    idempotency_key: str = Field(min_length=1)


class EvidenceSlotView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    evidence_slot_id: str
    evidence_role: str
    entity_scope: list[str]
    period_scope: str
    source_policy_ref: str
    required: bool


class DecisionSurfaceCellView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    cell_id: str
    cell_version: int = Field(ge=1)
    decision_question: str
    owner: str
    materiality: str
    stop_rule: str
    what_would_change: str
    evidence_slots: list[EvidenceSlotView]


class DecisionSurfaceView(BaseModel):
    model_config = ConfigDict(extra="forbid")

    case_id: str
    contract_id: str
    contract_version: int = Field(ge=1)
    contract_version_id: str
    checkpoint_version: int = Field(ge=1)
    review_status: Literal["awaiting_review", "accepted", "returned"]
    cells: list[DecisionSurfaceCellView]


def build_planning_router(service: PlanningService) -> APIRouter:
    router = APIRouter(tags=["point02-planning"])

    @router.post(
        "/cases/{case_id}/planning/compile",
        operation_id="compileDecisionSurface",
        status_code=status.HTTP_202_ACCEPTED,
        response_model=DecisionSurfaceView,
    )
    def compile_decision_surface(
        case_id: str,
        command: CompileDecisionSurfaceCommand,
        request: Request,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.compile_decision_surface(
                case_id,
                CompileDecisionSurfaceDraft(**command.model_dump()),
                _principal(tenant_id, project_id, actor_id, permissions),
                trace_id=request_trace_id(request),
            )
        except PlanningServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _planning_etag(projection)
        return projection

    @router.patch(
        "/cases/{case_id}/decision-surface",
        operation_id="reviseDecisionSurface",
        status_code=status.HTTP_202_ACCEPTED,
        response_model=DecisionSurfaceView,
    )
    def revise_decision_surface(
        case_id: str,
        command: ReviseDecisionSurfaceCommand,
        request: Request,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.revise_decision_surface(
                case_id,
                ReviseDecisionSurfaceDraft(
                    expected_case_version=command.expected_case_version,
                    expected_decision_surface_contract_version=command.expected_decision_surface_contract_version,
                    expected_checkpoint_version=command.expected_checkpoint_version,
                    changes=tuple(change.model_dump(exclude_none=True) for change in command.changes),
                    actor_ref=command.actor_ref,
                    idempotency_key=command.idempotency_key,
                ),
                _principal(tenant_id, project_id, actor_id, permissions),
                trace_id=request_trace_id(request),
            )
        except PlanningServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _planning_etag(projection)
        return projection

    @router.post(
        "/cases/{case_id}/planning/checkpoint",
        operation_id="reviewPlanningCheckpoint",
        status_code=status.HTTP_202_ACCEPTED,
        response_model=DecisionSurfaceView,
    )
    def review_planning_checkpoint(
        case_id: str,
        command: PlanningCheckpointDecisionCommand,
        request: Request,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.review_planning_checkpoint(
                case_id,
                PlanningCheckpointDecisionDraft(**command.model_dump()),
                _principal(tenant_id, project_id, actor_id, permissions),
                trace_id=request_trace_id(request),
            )
        except PlanningServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _planning_etag(projection)
        return projection

    @router.get(
        "/cases/{case_id}/decision-surface",
        operation_id="getDecisionSurface",
        response_model=DecisionSurfaceView,
    )
    def get_decision_surface(
        case_id: str,
        response: Response,
        tenant_id: Annotated[str | None, Header(alias="X-Fin-Case-Tenant")] = None,
        project_id: Annotated[str | None, Header(alias="X-Fin-Case-Project")] = None,
        actor_id: Annotated[str | None, Header(alias="X-Fin-Case-Actor")] = None,
        permissions: Annotated[str | None, Header(alias="X-Fin-Case-Permissions")] = None,
    ) -> dict[str, Any]:
        try:
            projection = service.get_decision_surface(
                case_id,
                _principal(tenant_id, project_id, actor_id, permissions),
            )
        except PlanningServiceError as exc:
            _raise_service_error(exc)
        response.headers["ETag"] = _planning_etag(projection)
        return projection

    return router


def _principal(
    tenant_id: str | None,
    project_id: str | None,
    actor_id: str | None,
    permissions: str | None,
) -> CasePrincipal:
    return CasePrincipal(
        tenant_id=(tenant_id or "").strip(),
        project_id=(project_id or "").strip(),
        actor_id=(actor_id or "").strip(),
        permissions=frozenset(item.strip() for item in (permissions or "").split(",") if item.strip()),
    )


def _raise_service_error(error: PlanningServiceError) -> None:
    raise HTTPException(status_code=error.status_code, detail=error.detail) from error


def _planning_etag(projection: dict[str, Any]) -> str:
    return (
        f'"decision-surface={projection["contract_version"]};'
        f'checkpoint={projection["checkpoint_version"]}"'
    )

File: archive/versions/pre_fin_0_1_3/unpromoted_active_tree/apps/workbench/frontend/vite/src/api/planning.ts
import { CASES_PATH, FixtureCaseContext } from "./cases";

export const P36_COMPILER_POLICY_REF = "fixture:p36-ten-cell-v1";
export const P36_PACK_SELECTION_REF = "fixture:p36-ai-infrastructure-v2";

export type CompileDecisionSurfaceCommand = {
  expected_case_version: number;
  expected_summary_version: number;
  compiler_policy_ref: typeof P36_COMPILER_POLICY_REF;
  pack_selection_ref: typeof P36_PACK_SELECTION_REF;
  actor_ref: string;
  idempotency_key: string;
};

export type DecisionSurfaceRevision = {
  cell_id: string;
  what_would_change: string;
  stop_rule?: string;
};

export type ReviseDecisionSurfaceCommand = {
  expected_case_version: number;
  expected_decision_surface_contract_version: number;
  expected_checkpoint_version: number;
  changes: DecisionSurfaceRevision[];
  actor_ref: string;
  idempotency_key: string;
};

export type PlanningCheckpointDecision = "accept" | "return";

export type PlanningCheckpointDecisionCommand = {
  decision: PlanningCheckpointDecision;
  expected_case_version: number;
  expected_decision_surface_contract_version: number;
  expected_checkpoint_version: number;
  actor_ref: string;
  idempotency_key: string;
};

export type EvidenceSlotView = {
  evidence_slot_id: string;
  evidence_role: string;
  entity_scope: string[];
  period_scope: string;
  source_policy_ref: string;
  required: boolean;
};

export type DecisionSurfaceCellView = {
  cell_id: string;
  cell_version: number;
  decision_question: string;
  owner: string;
  materiality: string;
  stop_rule: string;
  what_would_change: string;
  evidence_slots: EvidenceSlotView[];
};

export type DecisionSurfaceReviewStatus = "draft" | "awaiting_review" | "accepted" | "returned";

export type DecisionSurfaceView = {
  case_id: string;
  contract_id: string;
  contract_version: number;
  contract_version_id: string;
  checkpoint_version: number;
  review_status: DecisionSurfaceReviewStatus;
  cells: DecisionSurfaceCellView[];
};

type ErrorEnvelope = {
  error?: {
    error_code?: string;
    message?: string;
    status_code?: number;
    trace_id?: string;
  };
  detail?: unknown;
};

export class PlanningApiError extends Error {
  readonly code: string;
  readonly statusCode: number;
  readonly traceId: string | undefined;
  readonly detail: unknown;

  constructor(input: { code: string; message: string; statusCode: number; traceId?: string; detail?: unknown }) {
    super(input.message);
    this.name = "PlanningApiError";
    this.code = input.code;
    this.statusCode = input.statusCode;
    this.traceId = input.traceId;
    this.detail = input.detail;
  }
}

export const DEFAULT_FIXTURE_PLANNING_CONTEXT: FixtureCaseContext = {
  tenantId: "fixture_internal",
  projectId: "workbench_internal",
  actorId: "analyst_internal",
  permissions: ["case:read", "case:create", "planning:read", "planning:write", "planning:review"],
};

export class PlanningApiClient {
  constructor(private readonly context: FixtureCaseContext = DEFAULT_FIXTURE_PLANNING_CONTEXT) {}

  get actorRef(): string {
    return this.context.actorId;
  }

  async compileDecisionSurface(caseId: string, command: CompileDecisionSurfaceCommand): Promise<DecisionSurfaceView> {
    return this.request<DecisionSurfaceView>(`${casePath(caseId)}/planning/compile`, mutationInit("POST", command));
  }

  async getDecisionSurface(caseId: string): Promise<DecisionSurfaceView> {
    return this.request<DecisionSurfaceView>(`${casePath(caseId)}/decision-surface`);
  }

  async reviseDecisionSurface(caseId: string, command: ReviseDecisionSurfaceCommand): Promise<DecisionSurfaceView> {
    return this.request<DecisionSurfaceView>(`${casePath(caseId)}/decision-surface`, mutationInit("PATCH", command));
  }

  async reviewPlanningCheckpoint(caseId: string, command: PlanningCheckpointDecisionCommand): Promise<DecisionSurfaceView> {
    return this.request<DecisionSurfaceView>(`${casePath(caseId)}/planning/checkpoint`, mutationInit("POST", command));
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const headers = new Headers(init.headers);
    headers.set("Accept", "application/json");
    headers.set("X-Fin-Case-Tenant", this.context.tenantId);
    headers.set("X-Fin-Case-Project", this.context.projectId);
    headers.set("X-Fin-Case-Actor", this.context.actorId);
    headers.set("X-Fin-Case-Permissions", this.context.permissions.join(","));

    let response: Response;
    try {
      response = await fetch(path, { ...init, headers });
    } catch (error) {
      throw new PlanningApiError({
        code: "network_unavailable",
        message: "The fixture planning API is unavailable.",
        statusCode: 0,
        detail: error,
      });
    }

    const text = await response.text();
    const payload = text ? parsePayload(text) : {};
    if (!response.ok) {
      throw toPlanningApiError(response.status, payload);
    }
    return payload as T;
  }
}

function casePath(caseId: string): string {
  return `${CASES_PATH}/${encodeURIComponent(caseId)}`;
}

function mutationInit(method: "POST" | "PATCH", command: { idempotency_key: string }): RequestInit {
  return {
    method,
    headers: {
      "Content-Type": "application/json",
      "Idempotency-Key": command.idempotency_key,
    },
    body: JSON.stringify(command),
  };
}

function parsePayload(text: string): unknown {
  try {
    return JSON.parse(text) as unknown;
  } catch {
    return { detail: text };
  }
}

function toPlanningApiError(statusCode: number, payload: unknown): PlanningApiError {
  const envelope = isErrorEnvelope(payload) ? payload : {};
  const error = envelope.error;
  return new PlanningApiError({
    code: error?.error_code ?? `http_${statusCode}`,
    message: error?.message ?? detailMessage(envelope.detail) ?? "Planning request failed.",
    statusCode,
    traceId: error?.trace_id,
    detail: envelope.detail,
  });
}

function detailMessage(detail: unknown): string | undefined {
  if (typeof detail === "string") return detail;
  if (typeof detail !== "object" || detail === null) return undefined;
  const reason = (detail as { reason?: unknown }).reason;
  return typeof reason === "string" ? reason : undefined;
}

function isErrorEnvelope(value: unknown): value is ErrorEnvelope {
  return typeof value === "object" && value !== null;
}
